### PR TITLE
validate tutorials

### PR DESCRIPTION
**Pull request by amtep**

### CODE REVIEW

File: src/data/tutorials.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            vd.field_item("button_text", Item::Localization);
+            vd.field_validated_value("target", validate_lesson_target);
+            vd.field_validated_key_block("enabled", |key, block, data| {
+                let mut sc = ScopeContext::new(game_tutorial_scope(), key);
+                validate_trigger(block, data, &mut sc, Tooltipped::No);
+            });
+        });
+        vd.multi_field_validated_block("trigger_transition", validate_trigger_transition);
+
+        vd.field_validated_key_block("interface_effect", |key, block, data| {
```
Comment (my4ng) [inline src/data/tutorials.rs:154]: not present in vic3 doc... best to verify

File: src/data/tutorials.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+fn game_tutorial_scope() -> Scopes {
+    match Game::game() {
+        #[cfg(feature = "ck3")]
+        Game::Ck3 => Scopes::Character,
+        #[cfg(feature = "vic3")]
+        Game::Vic3 => Scopes::Country,
+        #[cfg(feature = "imperator")]
+        Game::Imperator => Scopes::Country,
```
Comment (my4ng) [inline src/data/tutorials.rs:203]: Shouldn't include imperator, or unreachable!()

**Review by my4ng — CHANGES_REQUESTED** [review]

File: src/data/tutorials.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        vd.field_validated_key_block("trigger", |key, block, data| {
+            let mut sc = ScopeContext::new(game_tutorial_scope(), key);
+            validate_trigger(block, data, &mut sc, Tooltipped::No);
+        });
+
+        // TODO: register these as item Flags and check them for
+        // the IsTutorialTagOpen gui function?
+        // Downside: some mods may check other mods' tags for compatibility.
+        vd.field_value("gui_tag");
```
[the remark this answers is not in the document]
Comment (amtep) [reply src/data/tutorials.rs:53]: Yep

File: src/data/tutorials.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        vd.field_value("highlight_widget");
+
+        vd.multi_field_validated_block("trigger_transition", validate_trigger_transition);
+
+        vd.field_integer("delay");
+        vd.field_integer("default_lesson_step_delay");
+
+        vd.field_bool("finish_gamestate_tutorial");
+        vd.field_bool("shown_in_encyclopedia");
+
```
[the remark this answers is not in the document]
Comment (amtep) [reply src/data/tutorials.rs:64]: Aha, so the Vic3 info is different :)

File: src/data/tutorials.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct TutorialLessonChain {
+    gamestate_tutorial: bool,
+}
+
+inventory::submit! {
+    ItemLoader::Normal(GameFlags::all(), Item::TutorialLessonChain, TutorialLessonChain::add)
```
[the remark this answers is not in the document]
Comment (amtep) [reply src/data/tutorials.rs:76]: done

File: src/data/tutorials.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amtep) [reply src/data/tutorials.rs:64]: They are used in LessonSteps too

File: src/data/tutorials.rs (hunk 1 above)
Comment (amtep) [reply src/data/tutorials.rs:154]: I added some TODO :)

File: src/data/tutorials.rs (hunk 2 above)
Comment (amtep) [reply src/data/tutorials.rs:203]: Ok

File: src/data/tutorials.rs [hunk 6]
```diff
@@ -50,9 +53,18 @@ impl DbKind for TutorialLesson {
         // TODO: register these as item Flags and check them for
         // the IsTutorialTagOpen gui function?
         // Downside: some mods may check other mods' tags for compatibility.
-        vd.field_value("gui_tag");
+        vd.multi_field_value("gui_tag");
         // TODO: check that this is a widget name
         vd.field_value("highlight_widget");
+        // TODO: verify this works in CK3 too
+        vd.field_validated_key("highlight_widget_dynamic_loc", |key, bv, data| {
+            validate_datatype_field(Datatype::Unknown, key, bv, data, false);
```
Comment (my4ng) [inline src/data/tutorials.rs:61]: should be `Datatype::CString`

File: src/data/tutorials.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (my4ng) [reply src/data/tutorials.rs:77]: I may have been wrong about this actually. It seems that these are only used in lesson steps

**Review by my4ng — CHANGES_REQUESTED** [review]

File: src/data/tutorials.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amtep) [reply src/data/tutorials.rs:77]: Yeah while they're only documented for Lessons. It's odd.

File: src/data/tutorials.rs (hunk 6 above)
Comment (amtep) [reply src/data/tutorials.rs:61]: Unfortunately probably not. The game engine will convert numbers into strings too when it's expecting a string. Tiger doesn't currently have a way to express this. It's one of the TODOs in the gui code :)

File: src/data/tutorials.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (my4ng) [reply src/data/tutorials.rs:77]: Yeah I think I am happy to leave it in both then, just in case

**Review by my4ng — APPROVED** [review]

**Pull request merged**